### ISSUE TITLE
Race in Makefile `depend`/`init` target

### ISSUE THREAD

**Issue by jtagcat** [istio/istio#48990]
I found a ~90% reproducible race when building without Docker.

What I was trying to do: Update [Arch istio](https://gitlab.archlinux.org/archlinux/packaging/packages/istio/) package (should be better named `istioctl`). `istioctl` doesn't actually need the dependencies as defined in Makefiles (depend/init) for building nor running it. Please correct me if I'm wrong, will create a new ticket for that later.

```
  export BUILD_WITH_CONTAINER=0
```

Makefile.core.mk:
```
# Downloads envoy, based on the SHA defined in the base pilot Dockerfile
init: $(TARGET_OUT)/istio_is_init init-ztunnel-rs
	@mkdir -p ${TARGET_OUT}/logs
	@mkdir -p ${TARGET_OUT}/release
```

`init` has two dependants, which are run **parallel**: `$(TARGET_OUT)/istio_is_init` (envoy) and `init-ztunnel-rs` (ztunnel).

bin/init.sh downloads envoy:
```
    # Download and extract the binary to the output directory.
    echo "Downloading ${SIDECAR}: $1 to $2"
    time ${DOWNLOAD_COMMAND} --header "${AUTH_HEADER:-}" "$1" | tar xz
# jtagcat ^^^^^^^^^^^^^^^^^^^^ This is curl
```

Tar errors the build:
```
tar: usr/local/bin: Cannot utime: No such file or directory
tar: usr/local: Cannot utime: No such file or directory
tar: usr: Cannot utime: No such file or directory
tar: Exiting with failure status due to previous errors
```

When you change the code to first write to a file, and then `cat file | tar xz`, everything works fine. The download takes ~5s and leaves a gap where tar constructs the directory structure, and while the binary part of the tar archive is downloaded.

Meanwhile in build_ztunnel.sh:
```
  time ${DOWNLOAD_COMMAND} --header "${AUTH_HEADER:-}" "$1" > "${BASE_NAME}"
# jtagcat ^^^^^^^^^^^^^^^^^^^ This is curl
  chmod +x "${BASE_NAME}"

  # Copy the extracted binary to the output location
  cp "${BASE_NAME}" "$2"
# jtagcat: no extraction took place!

  # Remove the extracted binary.
  rm -rf usr
# jtagcat: nor was usr created, the `--force` flag in rm does not cause `set -o errexit` to propagate errors. Even in envoy's case, the flag isn't needed.
```

The ztunnel routine completes faster, nuking `usr/`. Interestingly, it doesn't extract or create the directory, its download is not an archive, but a binary.

The script structure and comments indicate that the second script is a copy of the first with minor modifications. Mistakes happen. I guess I'll be submitting a patch.

**Comment by dhawton** [comment]
For the record: Prow runs these commands outside of docker (BUILD_WITH_CONTAINER=0), so in theory these should be caught by our tests if it is reproducible. It could be just a difference between build environments, I don't believe our make targets run in parallel in our CI.
